File: bitlap-core/src/main/kotlin/org/bitlap/core/sql/BitlapSqlValidator.kt
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.core.sql

import org.apache.calcite.config.CalciteConnectionConfig
import org.apache.calcite.prepare.CalciteCatalogReader
import org.apache.calcite.rel.type.RelDataType
import org.apache.calcite.sql.SqlNode
import org.apache.calcite.sql.SqlOperatorTable
import org.apache.calcite.sql.SqlSelect
import org.apache.calcite.sql.validate.SqlValidator
import org.apache.calcite.sql.validate.SqlValidatorImpl
import org.bitlap.common.exception.BitlapException
import org.bitlap.core.sql.udf.FunctionRegistry
import org.bitlap.core.sql.udf.UdfNames

/**
 * Mail: chk19940609@gmail.com
 * Created by IceMimosa
 * Date: 2021/9/10
 */
class BitlapSqlValidator(
    val opTab: SqlOperatorTable,
    val catalogReader: CalciteCatalogReader,
    val config: SqlValidator.Config,
    val connConfig: CalciteConnectionConfig,
) : SqlValidatorImpl(opTab, catalogReader, catalogReader.typeFactory, config) {

    private val queryContext = QueryContext.get()

    override fun validate(topNode: SqlNode): SqlNode {
        return super.validate(topNode)
    }

    override fun validateSelect(select: SqlSelect, targetRowType: RelDataType) {
        queryContext.currentSelectNode = select
        val check = FunctionRegistry.sqlValidatorFunctions().invoke().map { it.validate(select, targetRowType) }
        if (!check.all { it }) {
            throw BitlapException("Incorrect syntax, ${UdfNames.date_format} should be also in group by.")
        }
        super.validateSelect(select, targetRowType)
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/BitlapSqlValidator.scala
/*
 * Copyright 2020-2023 IceMimosa, jxnu-liguobin and the Bitlap Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.bitlap.core.sql

import org.apache.calcite.config.CalciteConnectionConfig
import org.apache.calcite.prepare.CalciteCatalogReader
import org.apache.calcite.rel.`type`.RelDataType
import org.apache.calcite.sql.SqlNode
import org.apache.calcite.sql.SqlOperatorTable
import org.apache.calcite.sql.SqlSelect
import org.apache.calcite.sql.validate.SqlValidator
import org.apache.calcite.sql.validate.SqlValidatorImpl

class BitlapSqlValidator(
  val _opTab: SqlOperatorTable,
  val _catalogReader: CalciteCatalogReader,
  val _config: SqlValidator.Config,
  val connConfig: CalciteConnectionConfig)
    extends SqlValidatorImpl(_opTab, _catalogReader, _catalogReader.getTypeFactory, _config) {

  private val queryContext = QueryContext.get()

  override def validate(topNode: SqlNode): SqlNode = {
    super.validate(topNode)
  }

  override def validateSelect(select: SqlSelect, targetRowType: RelDataType): Unit = {
    queryContext.currentSelectNode = select
    super.validateSelect(select, targetRowType)
  }
}
